### PR TITLE
Smoothed pressure advance and extruder cornering support

### PR DESCRIPTION
**Pull request by KevinOConnor**
This PR significantly changes the internals of pressure advance in the hope that it will make it more flexible and scalable.  It also adds support for the extruder to make "instantaneous velocity" changes during cornering between moves.  This allows Klipper to perform lookahead between moves with different extrusion ratios.

The code is still a "work in progress".  The main goal at this point is to enable testing and to "prove the concept".  To test with this new code, run:

cd ~/klipper ; git fetch ; git checkout origin/work-smoothpa-20190920 ; sudo service klipper restart

It may be necessary to recalibrate pressure advance.  See the [updated guide on the new branch](https://github.com/KevinOConnor/klipper/blob/work-smoothpa-20190920/docs/Pressure_Advance.md).  This guide also introduces a new pressure advance tuning method (it uses a "tuning tower").

It is hoped that this new pressure advance system will be significantly more scalable.  The old code determined the amount of filament needed to account for pressure by calculating the nominal instantaneous velocity of the extruder.  This new code updates that calculation to use an average velocity over a small time window.  Using an average velocity avoids pressure spikes from obscure corner cases, it allows the extruder to make instantaneous velocity changes, it provides more time for pressure advance to build pressure (which should allow for greater pressure advance values), and it allows pressure advance to start building pressure in advance of an extrusion move.  The average velocity calculation will automatically span multiple g-code moves as needed.

See the updated example.cfg file for further details on tuning parameters.

As above, this code is a work in progress.  Some known caveats:
* The code will not work well with force_move and stepper_buzz.  Avoid using these commands on the extruder motor.
* The documentation has not yet been updated (kinematics.md, code_overview.md).
* Probably other bugs.

If you run tests with this code, please post your results here (success, failure, or other).

-Kevin

EDIT: Updates as new bugs found and fixed.

### CODE REVIEW

**Comment by tonn0297** [comment]
im having some weird issues.. when i print the calibration cube the layers are perfect. but if i then go and print a 30x30 cube my layers are inconsistent and the corneres are too "smooth"

some pic for refrence

my calibration the mark is at 0.72
https://media.discordapp.net/attachments/431558350833188894/625357981248126976/unknown.png

test with a cube 
https://cdn.discordapp.com/attachments/431558350833188894/625365812701822978/IMG_20190922_182000.jpg

test with diffrent pa
https://cdn.discordapp.com/attachments/431559280018194433/625394053487394844/unknown.png

test at 0.5 pa
https://cdn.discordapp.com/attachments/431559280018194433/625401671085981696/IMG_20190922_204234_1.jpg

ping me if you need me to test stuff or alter somthing

-Tonny

**Comment by KevinOConnor** [comment]
@Hywelmartin - thanks.

> it works, but some information is missing for tuning the time

I was working on the tuning test for pressure_advance_smooth_time when I ran into the conflict with homing (mentioned in the caveats section).  I need to track that down before completing the tuning guide.  In general, a higher smooth time should permit higher pressure advance values with high acceleration.

> and smooth time 0.002

Interesting that you've reduced the pressure_advance_smooth_time to .002 (from the default of .020).  What differences did you experience with the lower smooth time?

> should be
`pressure_advance = <start> + <measured_height> * <increment>`. (For example, 0.000 + 12.90 * .020 would be .258.)

Oops.  Thanks.  Fixed.

> i see it as the settings in printer.cfg is max limits

FYI, the pressure_advance setting in printer.cfg does not enforce a maximum limit.  The SET_PRESSURE_ADVANCE command may set any value irrespective of the config file.  The config file is just the default.

-Kevin

**Comment by Hywelmartin** [comment]
Opus.. the smooth time was unaltered(.020)... to many zeros.. maybe the setting should be in microseconds  =)

**Comment by Hywelmartin** [comment]
update:
I fiddled a bit with smooth_time..
It is rather difficult to find a sweet spot as they depend on each other in parallel, higher PA seems to need higher Smooth_time for the same-ish results.
0(zero) as a starting point does not work.. spaghetti mayhem.. probably due to skipped steps of the extruder...
I started at 0.005 and used the increments of 0.0005
`{if layer_z == 0.6}TUNE_PRESSURE_ADVANCE_SMOOTH_TIME START=0.005 INCREMENT=0.0005{endif}
{if layer_z == 0.6}TUNE_PRESSURE_ADVANCE START=0.05 INCREMENT=.005{endif}`
and ended up with 0.0135 at PA 0.22, unfortunately i destroyed  the print.. so no picture..

I then ran the PA test again and ended up at 0.215...
bild
_this is the start/stop corner_
-The arrow at the top is PA 0.215 level
-The arrow at the bottom is where the travel move starts after wipe
-Could you make the base circle a little bit lower, so that it could be printed with one bottom layer only.. it's much simpler to split the print and inspect the inside with only one bottom layer instead of 3..

Then a speed test..
Base: speed 200 acc 5000 acc_to_decel 3500 Cornerspeed 25.. I know that the acc/acc_to_deccel crippled the top speed...
bild
_If i remember correct.._
At 250% I increased the cooling..and the tension of the extruder.. and it helped..but as you see at the next picture there is a lack of plastic anyway..
the needed amount of PA seems to decrease as the speed gets higher. 
Due to altered viscosity?? 
Higher friction??
I normally print rather hot 267⁰C for ABS..(not tested the accuracy of my PT100) and I have a volcano with a 10mm extension on top..
The errors is much more pronounced during decelleration(or only, I haven't decided yet..) 
bild
_Slic3r/prusaslicer prints CCW as seen from above_

What is the "gcode" for altering the smooth_time if there is any...

A little [movie](https://youtu.be/MnoW1OaJyD0).. PA~0.5 Smooth_Time 0.020

**Comment by KevinOConnor** [comment]
Thanks @Hywelmartin.

I'm not sure I understand the results of your tests.  Could you summarize your findings?

> Base: speed 200 acc 5000 acc_to_decel 3500 Cornerspeed 25..

Those settings are quite high - I think that may be asking too much from pressure advance.  (With high speed, high pa, and high acceleration, the extruder has very little time to move the necessary filament - the extruder's torque and filament grip become a bottleneck and likely skew results.)

-Kevin

**Comment by KevinOConnor** [comment]
FYI, the homing issue with pressure_advance_smooth_time should hopefully now be fixed.  For anyone interested in testing pressure_advance_smooth_time, I use the following procedure:

1 - find the appropriate pressure_advance value using the tuning tower described in the docs
2 - run RESTART to return to normal printing accelerations and settings
3 - run `TUNE_PRESSURE_ADVANCE_SMOOTH_TIME START=.55 INCREMENT=-.004`
4 - print the tuning tower
5 - inspect the tuning tower to see the impact of different smooth time settings

FWIW, I haven't noticed an overall improvement to tuning the smooth time in my tests so far.  A higher smooth time definitely reduces the amount of extruder movement and I'm sure it would enable higher overall accelerations.  However, I found it resulted in noticeable blemishes on the wall leading up to corners (the earlier retractions caused by the higher smooth time resulted in a noticeable filament disturbance).

-Kevin

**Comment by Hywelmartin** [comment]
> Thanks @Hywelmartin.
> 
> I'm not sure I understand the results of your tests. Could you summarize your findings?
> 
> > Base: speed 200 acc 5000 acc_to_decel 3500 Cornerspeed 25..
> 
> and so on...

..In my rambling... from both posts.. 
I find it easier to find a robust setting if i run close or over the limit and then go back.
Findings
- Activate the test procedural(TUNE_PRESSURE_ADVAN...) during the first layer, adding it to the start of the gcode may not work..
-Smooth_time and PA work in parallel. I see Smooth time as a secondary tuning parameter(0.02 would work for most i think)
-Start of the tuning smooth_time shouldn't be from 0(zero)
-It is hard to find a  good value(at my aggressive speed and acc setting)..
-The errors(lack of filament) are mostly into the corner.
-corner_speed plays a big role when printing organic(splines) shapes.. and using PA...

**Conclusion..**
The test procedure works GOOD..as long as you don't start the time at 0(zero)
Smooth_time could be left at standard value, but there is a possibility to use lower time and less PA, with more aggressive movements of the extruder.
I have never had this sharp corners.

**What I ended up with**

[photos](https://photos.app.goo.gl/JrFuALA1Vzw9FGJm9)

[video](https://youtu.be/Su3vA7i4iTE)

setting for the lowest part(~3mm) was the ones from the testing procedure and it's a keeper..
settings common for the whole print..
-`SET_PRESSURE_ADVANCE ADVANCE=0.215  SMOOTH_TIME=0.0135`
-`SET_VELOCITY_LIMIT ACCEL=8000 ACCEL_TO_DECEL=1500 SQUARE_CORNER_VELOCITY=9`
-Layer height 0.2
-Speeds
image

Link for STL
https://www.dropbox.com/s/b81px3crksxaq9g/PA_test.stl?dl=0

**Comment by EtteHub** [comment]
Still in test with this new PA, but I have notice something:

The text states 'The TUNE_PRESSURE_ADVANCE command instructs Klipper to alter the pressure_advance setting on each layer of the print' meaning (or at least implying) the increment is layer related.

But your formula uses the height, not the layer number (which implies an increment per mm).

Which implementation is the real one?

**Comment by cosminr86** [comment]
because is more easy to measure a distance and divide by layer height than counting number of layers

**Comment by EtteHub** [comment]
Let me rephrase what I said:

Is the increment every layer or every mm? because text and formula are not 'saying' the same thing

**Comment by cosminr86** [comment]
every layer

**Comment by Hywelmartin** [comment]
The text is right.. Klipper alter the value every layer, not by the the increment, but according to hight.

So increment is value/mm

**Comment by EtteHub** [comment]
Oh ok, thanks for the clarifications

**Comment by KevinOConnor** [comment]
FYI, I've updated the tuning tower commands (there's now a single generic TUNING_TOWER command).  Overall functionality should not have changed.

-Kevin

**Comment by cosminr86** [comment]
I had the same problem, tuning script is not working if it is executed before print start

**Comment by KevinOConnor** [comment]
> I had the same problem, tuning script is not working if it is executed before print start

I think this should be fixed on the latest version of the code.

-Kevin

**Comment by KevinOConnor** [comment]
FYI, I've done a bunch of further tests.  I've concluded that 20ms for pressure_advance_smooth_time (20ms) is not a good default.  It works fine on my delta printer (with PA=.350), but on my printer with a direct drive extruder (PA=.050) the default leads to noticeable blemishes on the wall leading up to the corner during prints with normal acceleration.  On that printer, it seems values closer to 5ms are needed to eliminate blemishes. I suspect there may be an ideal value of something like `smooth_time = .5 * pressure_advance`.

I also have a general plan to see if an alternate implementation can reduce the impact of larger smooth time settings.

-Kevin

**Comment by jalanjarosz** [comment]
@KevinOConnor 

I’ve been following this thread and since I’ve been struggling with PA on my CoreXY, I’m curious if this test code may help?

CoreXY Settings:
max_velocity: 150
max_accel: 1500
square_corner_velocity: 5

Extruder:
filament_dia: 1.75
max_extrude_only_velocity: 120
max_extrude_only_accel: 150
pressure_advance: 0.5

BMG Clone extruder w/ 3:1 ratio
850mm long Bowden 

If you think it would, what starting point should I use?

DCD4DFB1-C698-4A16-975B-A9B8494EF7BA
200B9A26-396C-4D9C-AB8B-F8FAAB4D3C41

**Comment by wasalm** [comment]
I have an Ender 3 with direct drive and a volcano clone. I tried this branch with a 3dBenchy and it works great.

**Comment by KevinOConnor** [comment]
@jalanjarosz - I don't know of a way to tune pressure advance from an arbitrary print - if you're struggling with configuration of pressure_advance, be sure to follow the pressure advance tuning guide on the master branch (or you can try the tuning guide on this development branch).  If you run into problems using the tuning guide then you can try attaching pictures of those tuning prints.

If you think you're running into an issue with gap fill (as in #1758) or one of the slicer conflicts described in #1869 then this branch may help.  See those issues for the details.  If you do run a test on this branch, lets us know the results.

-Kevin

**Comment by JohnEdwa** [comment]
I've been testing this for a few prints now, and while it looks and sounds a lot better, and prints noticeably faster than it ever did, the overall quality has gotten worse. Not because the changes are bad, but because it's pushing my printer harder than it used to - the old implementation was clearly really slowing it down in some places.

I still have plenty of settings to tweak and I will no doubt get the results I'm looking for, but my hunch is that this is a change that might result in many people having to re-tune their printers and should probably be implemented with the next major version bump and not just "silently" merged into the code.

**Comment by gaze** [comment]
This is a really cool update! I am however running into a funny issue where I see these defects sometimes going into and sometimes going away from corners. They don't appear to be ringing. Running a PA of 0.6. 
highPA

**Comment by KevinOConnor** [comment]
@gaze - I suspect it's the `pressure_advance_smooth_time` as mentioned here: https://github.com/KevinOConnor/klipper/pull/1997#issuecomment-536366576

You could try printing the tuning tower while adjusting the smooth time parameter.  Use normal printing speed settings, but run this command prior to printing the tuning tower: `TUNING_TOWER COMMAND=SET_PRESSURE_ADVANCE PARAMETER=SMOOTH_TIME START=101 FACTOR=-.002`

-Kevin

**Comment by gaze** [comment]
Both are the tuning towers run with

`TUNING_TOWER COMMAND=SET_PRESSURE_ADVANCE PARAMETER=SMOOTH_TIME START=101 FACTOR=-.002`

Left is with 2500 mm/s^2, right is with 2000 mm/s^2. Seems like you shouldn't go too high in acceleration but you also shouldn't go too low?

image0

**Comment by KevinOConnor** [comment]
I've updated this branch to use a new "balanced smoothing time".  Previously, the amount of pressure advance needed at time t was calculated based on the average velocity between time `t` and `t + smooth_time`.  Now the code calculates the average between time `t - smooth_time` and `t + smooth_time`.

I'm finding much better behaviour with this change.  I'm seeing less side-wall defects and I can go up to higher pressure_advance_smooth_time values.

It may be necessary to recalibrate after pulling the latest code.

-Kevin

**Comment by EtteHub** [comment]
Just got an error in a middle of a 6H print : MCU disconnected, Timer too close

FYI very last version (with the balanced smoothing time) and the part it stopped on was with very small pillars next to each other.

I will switch to the regular master and restart the print. In case the same error appears again (i.e. not related to this fork modifications) I'll let you know. It is the first time I'm facing such an error.

**Comment by nophead** [comment]
Why is timer too close a fatal error. Most systems would just deliver a late pulse and continue. That seems a lot better than trashing a 6 hour print.

**Comment by KevinOConnor** [comment]
A "timer too close" error is indicative of a communications failure between host and micro-controller.  (The typical cause is the host sending a command to the micro-controller that schedules an event in the past.)  Without the full log there isn't much we can do.  It seems unlikely to be related to this pressure advance branch.

-Kevin

**Comment by ambethia** [comment]
@KevinOConnor 

`TUNING_TOWER COMMAND=SET_PRESSURE_ADVANCE PARAMETER=SMOOTH_TIME START=101 FACTOR=-.002`

101 seems like an odd start value, am I misunderstanding something?

**Comment by KevinOConnor** [comment]
> 101 seems like an odd start value, am I misunderstanding something?

I choose that value so that a 50mm tuning tower wouldn't result in a negative smooth time at the top of the print.  More generally though, all the calibration instructions here should be thought of as "starting points" - everything on this branch is still experimental.  There are more details on the TUNING_TOWER command in docs/G-Codes.md.

-Kevin

**Comment by ambethia** [comment]
Ahh, Kevin, thank you. I didn't notice the negative value for the factor. It makes sense now.

**Comment by LaurenceGGush** [comment]
Hi Kevin thanks for your continued hard work on klipper, really awesome.

I've being trying out this branch and really liking the results, I can use a lot more advance on bowden printer without skipping steps.

That said I have run into a "timer too close" issue on the latest "balanced smooth time" version.
I appears to be caused by the very small moves cura produces to fill gaps, see screenshot. If I turn off "Fill Gaps Between Walls" I don't get the error.
If I go back to master I don't get the error.

[klippy-smooth-pa-timer-too-close.log](https://github.com/KevinOConnor/klipper/files/3694500/klippy-smooth-pa-timer-too-close.log)
cura-smooth-pa-timer-too-close

**Comment by KevinOConnor** [comment]
@LaurenceGGush - thanks.  At first glance it does appear to be a host scheduling regression introduced by this branch.  My initial interpretation is that OctoPrint could not send g-codes fast enough, Klipper transitioned to a "print stall" state, but the new code on this branch failed to flush its buffers properly.  I suspect a print from virtual_sdcard would not have the issue.

I'll try to take a closer look to track it down.

-Kevin

**Comment by EtteHub** [comment]
> @gaze - I suspect it's the `pressure_advance_smooth_time` as mentioned here: [#1997 (comment)](https://github.com/KevinOConnor/klipper/pull/1997#issuecomment-536366576)
> 
> You could try printing the tuning tower while adjusting the smooth time parameter. Use normal printing speed settings, but run this command prior to printing the tuning tower: `TUNING_TOWER COMMAND=SET_PRESSURE_ADVANCE PARAMETER=SMOOTH_TIME START=101 FACTOR=-.002`
> 
> -Kevin

I think you have a typo and the SMOOTH_TIME START should be at 0.101 and not 101 (50x0.002 = 0.1)

**Comment by glowtape** [comment]
Is the smooth time still supposed to be 20ms for this, now that the latest version also looks into both directions?

**Comment by KevinOConnor** [comment]
I was finding good results with a total smooth time of 40ms.  I may end up going with a total smooth time of 30ms (or some other value depending on feedback).  Currently the `pressure_advance_smooth_time` parameter is confusing (it's specifying half the total smooth time) - that's one of the many things needing to be fixed before final commit.

-Kevin

**Comment by RustyWraith** [comment]
I've got "timer too close" error. I believe, it occurred on printing single wall extrusion right on top of a part.
image

**Comment by drelich** [comment]
I must be doing something wrong, haha. I can only see marginal improvements on my test tower but I can confirm the settings were changing during the print. I used the default settings for a bowden type printer (Ender 3 with capricorn xs tube but fully stock hotend):

`SET_VELOCITY_LIMIT SQUARE_CORNER_VELOCITY=1 ACCEL=500`
followed by
`TUNING_TOWER COMMAND=SET_PRESSURE_ADVANCE PARAMETER=ADVANCE START=0 FACTOR=.020`

And here are the photos. Either I am going blind (which wouldn't surprise me lol) or the effect really is barely noticeable? I do see the 'line' coming closer to the corner towards the top but the corners are the same 'bulky' ones all over the print. Anybody can recommend other settings that might prove more helpful? Cheers.

IMG_0164
IMG_0165
IMG_0166
Screenshot 2019-10-10 at 14 39 11

**Comment by KevinOConnor** [comment]
@drelich - Make sure you're on the latest code on this branch.  I'd try altering the TUNING_TOWER parameters and retry - for example: `TUNING_TOWER COMMAND=SET_PRESSURE_ADVANCE PARAMETER=ADVANCE START=0.500 FACTOR=.040`.  If you find the extruder starts skipping, you can further reduce the acceleration during the test.

Note, however, when you do find the final pressure advance value, it may be too high to be practical during normal prints.  In which case, you may wish to not use pressure advance.  Alternatively, some people make hardware modifications (eg, reducing the length of the bowden tube) to improve the extruder response.

-Kevin

**Comment by drelich** [comment]
Thanks, @KevinOConnor !

I do have the bowden tube slightly shorter compared to the stock one so it might affect the results. 

I'll try as you recommend with the updated TUNING_TOWER parameters and report back.

David

**Comment by KevinOConnor** [comment]
I've updated this branch with the intent of getting it ready to merge.  It is now based on top of the code in PR #2130.  I think the "timer too close" error should now also be fixed.

In addition to the rework, this code now applies a general "positional smoother" to the extruder when pressure advance is enabled.  It is hoped this will help further smooth out the extruder movement.  It may be necessary to retune pressure advance as a result.

Also, the default for pressure_advance_smooth_time is now 40ms - which is actually the same as the previous default (20ms before actually smoothed over 40ms).

-Kevin

**Comment by glowtape** [comment]
So I let the printer do its thing while at work, just like yesterday, the same print, except with this branch and specific PA values (same Gcode file, except a different SET_PRESSURE_ADVANCE parameters). 

When I came home (to a finished print), Klipper was spewing this into the terminal window in OctoKlipper over and over, every M105 that Octoprint issued:

```
Send: M105
Recv: ok B:19.0 /0.0 T0:19.5 /0.0
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Send: M105
Recv: ok B:19.0 /0.0 T0:19.3 /0.0
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
Recv: !! Must home axis first: 125.000 245.000 41.360 [24897.550]
...
```

That seems new, it didn't do that before checking out the branch.

In klippy.log it looked somewhat like this, repeatedly:

```
EndstopError: Must home axis first: 125.000 245.000 41.360 [24897.550]
Must home axis first: 125.000 245.000 41.360 [24897.550]
idle timeout gcode execution
Traceback (most recent call last):
  File "/home/pi/klipper/klippy/extras/idle_timeout.py", line 45, in transition_idle_state
    res = self.gcode.run_script(script)
  File "/home/pi/klipper/klippy/gcode.py", line 292, in run_script
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/pi/klipper/klippy/gcode.py", line 224, in _process_commands
    handler(params)
  File "/home/pi/klipper/klippy/gcode.py", line 74, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/pi/klipper/klippy/gcode.py", line 513, in cmd_G1
    self.move_with_transform(self.last_position, self.speed)
  File "/home/pi/klipper/klippy/extras/bed_mesh.py", line 160, in move
    self.toolhead.move([x, y, z + self.fade_target, e], speed)
  File "/home/pi/klipper/klippy/toolhead.py", line 391, in move
    self.kin.check_move(move)
  File "/home/pi/klipper/klippy/kinematics/corexy.py", line 85, in check_move
    self._check_endstops(move)
  File "/home/pi/klipper/klippy/kinematics/corexy.py", line 78, in _check_endstops
    end_pos, "Must home axis first")
EndstopError: Must home axis first: 125.000 245.000 41.360 [24897.550]
```

After issuing QUERY_ENDSTOPS on a hunch based on the _check_endstops() call in the trace, it shut up and just merrily polled temps without further errors.

I issued M84 remotely after I noticed the print was done and the printer cooling down, I'm not sure if it's related. I can't tell if it was triggered by that from klippy.log.

**Comment by glowtape** [comment]
There's one very tiny layer shift in the X direction only (maybe 0.25mm-ish), which seem odd, because the printer's a CoreXY. One stepper crapping out for a short bit would shift it diagonally. I'd figure the same would happen when a nozzle gets stuck (altho there's nothing special on the affected layer that looks like it could be problematic, it's just more of the same).

Not sure if it might be related to this PR. I need another copy of that part, it should probably happen again printing the exact same file.

**Comment by KevinOConnor** [comment]
> I issued M84 remotely after I noticed the print was done and the printer cooling down, I'm not sure if it's related.

I suspect that's the likely cause.  It seems you have an idle_timeout script with movement - that would cause the idle_timeout script to fail - which causes it to be rerun once a second.  So, you're just getting the errors from that - any command that causes any "printer activity" will reset the idle_timeout, so the errors stop (at least until the timeout fires again.)

-Kevin

**Comment by glowtape** [comment]
Oh there's indeed one. Never noticed it. The config came from a template.

```
[idle_timeout]
gcode:
    G0 X125 Y245 F6000
    M84
timeout: 3600
```

Thanks for pointing it out!

**Comment by glowtape** [comment]
The layer shift was a false alarm, must have been some mechanical issue. Same Gcode printed fine today.

**Comment by KevinOConnor** [comment]
FYI, I've rebased this branch a few times now - mostly tweaking the host timing.  The most recent rebase was this morning.  (There hasn't been any kinematic changes since https://github.com/KevinOConnor/klipper/pull/1997#issuecomment-549624572 though.)

At this point, the additional homing/probing delays should be mostly eliminated now.

Assuming no further issues, I have a rough plan to merge in the next few days.

-Kevin

**Comment by glowtape** [comment]
Did a bunch of prints on the latest rebase, no oddities to report.
